### ISSUE TITLE
BUG - repeated record in input file is not read deterministically

### ISSUE THREAD

**Issue by Paulie14** [flow123d/flow123d#1219]
**Describe the bug**
See issue #1199 for details and data files.

The problem is however somehow hidden in the yaml input file.
In both input files there are two balance records, which should definitely be avoided! 
```
balance: {}
balance:
  cumulative: true
```

The first record means to create Balance object without the cumulative data, the second otherwise.
If I used only one the records, it worked correctly as expected.

However I was not able to explain, when it has both records at the same time, why one of the file computed the cumulative values and the other did not.

For devs
balance.cc:147 ```cumulative_ = in_rec.val<bool>("cumulative");```
gives true in one case and false in the other. I do not know why since the difference between the input files is only due to output of `piezo_head_p0`.

**Comment by Paulie14** [comment]
Yaml specifies the duplicates as an error:
https://yaml.org/spec/1.2/spec.html#style/single%20key:value%20pair%20mapping/

Similar problem reported for `yaml-cpp` here:
https://stackoverflow.com/questions/56057851/in-which-order-yaml-cpp-returns-data

The order of nodes is unspecified (chosen internally).
Own words of the author of `yaml-cpp` on duplicates:
> yaml-cpp is simply lenient here and doesn't report an error.

So I learnt that 'lenient' means 'shovívavý' in Czech. This sucks...

**Comment by jbrezmorf** [comment]
After digging a while in the yaml-cpp sources I realized that fixing that would be quite nontrivial because of many layers of the code and since maps are not stored as maps but as a list of (key, value) pairs. So it is time to try another library:

[rapidyaml](https://github.com/biojppm/rapidyaml) 
... also list of (key, value) pairs 

or

[libyaml](https://github.com/yaml/libyaml) ... use list of (key, value) pairs as well.

We should try if these have better checks.

**Comment by jbrezmorf** [comment]
Key equivalence in YAML can be application specific  that's the reason why all libraries have a similar approach. 
We have to do the check ourself:
1. modify PathYAML::down methods and the constructor to include check_map_unique check for resulting node.  
2. implement check_map_unique to 
   - check that node is_map
   - iterate over all items using YAML::Node::begin() ... end() which iterates over (key, Node) pairs.
     put the keys into a vector V and use:
     ```
     end = V.end();
     std::sort(V.begin(), end);
    bool containsDuplicates = (std::unique(V.begin(), end) != end);
    ```

**Comment by Paulie14** [comment]
Solved by d3d8d81394752f7941f4f5f6a45ff2e9816a397b.
